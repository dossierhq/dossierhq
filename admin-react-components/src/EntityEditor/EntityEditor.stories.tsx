import { ok } from '@datadata/core';
import type { Story } from '@storybook/react/types-6-0';
import React from 'react';
import { DataDataContext } from '../..';
import { EntityEditor } from './EntityEditor';
import type { EntityEditorProps } from './EntityEditor';
import schema from '../StoryboardSchema';

export default {
  title: 'Components/EntityEditor',
  component: EntityEditor,
  argTypes: { onSubmit: { action: 'submit' } },
  args: {},
};

const Template: Story<EntityEditorProps> = (args) => {
  return (
    <DataDataContext.Provider value={{ schema, searchEntities: () => Promise.resolve(ok(null)) }}>
      <EntityEditor {...args} />
    </DataDataContext.Provider>
  );
};

export const NewFoo = Template.bind({});
NewFoo.args = { entity: { _type: 'Foo' }, idPrefix: 'new-entity-123' };

export const FullFoo = Template.bind({});
FullFoo.args = {
  entity: {
    id: 'fc66b4d7-61ff-44d4-8f68-cb7f526df046',
    _type: 'Foo',
    _name: 'Hello',
    _version: 0,
    title: 'Hello',
  },
};

export const DeletedFoo = Template.bind({});
DeletedFoo.args = {
  entity: {
    id: 'fc66b4d7-61ff-44d4-8f68-cb7f526df046',
    _type: 'Foo',
    _name: 'Hello',
    _version: 1,
    _deleted: true,
  },
};
