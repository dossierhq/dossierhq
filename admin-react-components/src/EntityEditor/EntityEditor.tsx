import type {
  AdminEntity,
  AdminEntityCreate,
  AdminEntityUpdate,
  EntityTypeSpecification,
  FieldSpecification,
  Schema,
} from '@datadata/core';
import React, { useState } from 'react';
import { Divider, EntityFieldEditor, Form, FormField, InputSubmit, InputText } from '..';

interface NewEntity {
  _type: string;
}

export interface EntityEditorProps {
  idPrefix?: string;
  entity: NewEntity | AdminEntity;
  onSubmit: (entity: AdminEntityCreate | AdminEntityUpdate) => void;
  schema: Schema;
}

interface EntityEditorState {
  name: string;
  initialName: string;
  fields: FieldEditorState[];
}

interface FieldEditorState {
  fieldSpec: FieldSpecification;
  value: unknown;
  initialValue: unknown;
}

export function EntityEditor({
  idPrefix,
  entity,
  schema,
  onSubmit,
}: EntityEditorProps): JSX.Element {
  const { _type: type } = entity;
  const entitySpec = schema.getEntityTypeSpecification(type);
  if (!entitySpec) {
    throw new Error(`No such entity type in schema (${type})`);
  }

  const [state, setState] = useState(() => createInitialState(entitySpec, entity));
  const [resolvedIdPrefix] = useState(
    idPrefix
      ? idPrefix
      : 'id' in entity
      ? `entity-${entity.id}`
      : `new-entity-${String(Math.random()).slice(2)}`
  );

  return (
    <Form onSubmit={() => onSubmit(createAdminEntity(entity, state))}>
      <FormField
        controlId={`${resolvedIdPrefix}-_name`}
        label="Name"
        render={({ id }) => (
          <InputText id={id} value={state.name} onChange={(x) => setState({ ...state, name: x })} />
        )}
      />
      <Divider />
      {state.fields.map(({ fieldSpec, value, initialValue }, index) => {
        const handleFieldChanged = (newValue: unknown) => {
          const newFields = [...state.fields];
          newFields[index] = { fieldSpec, value: newValue, initialValue };
          setState({ ...state, fields: newFields });
        };

        return (
          <EntityFieldEditor
            idPrefix={resolvedIdPrefix}
            key={fieldSpec.name}
            fieldSpec={fieldSpec}
            value={value}
            onValueChanged={handleFieldChanged}
          />
        );
      })}

      <InputSubmit value="Save" disabled={!state.name} />
    </Form>
  );
}

function createInitialState(
  entitySpec: EntityTypeSpecification,
  entity: NewEntity | AdminEntity
): EntityEditorState {
  const adminEntity = 'id' in entity ? entity : null;
  return {
    name: adminEntity?._name || '',
    initialName: adminEntity?._name || '',
    fields: entitySpec.fields.map((fieldSpec) => {
      const value = adminEntity?.[fieldSpec.name] ?? null;
      return { fieldSpec, value, initialValue: value };
    }),
  };
}

function createAdminEntity(
  originalEntity: NewEntity | AdminEntity,
  state: EntityEditorState
): AdminEntityCreate | AdminEntityUpdate {
  let result: AdminEntityCreate | AdminEntityUpdate;
  if ('id' in originalEntity) {
    result = { id: originalEntity.id, _type: originalEntity._type };
    if (state.name !== state.initialName) {
      result._name = state.name;
    }
  } else {
    result = { _type: originalEntity._type, _name: state.name };
  }
  for (const { fieldSpec, value, initialValue } of state.fields) {
    if (value !== initialValue) {
      result[fieldSpec.name] = value;
    }
  }
  return result;
}
