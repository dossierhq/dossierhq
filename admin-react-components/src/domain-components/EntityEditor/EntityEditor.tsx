import type { AdminEntityCreate, AdminEntityUpdate } from '@datadata/core';
import type { Dispatch, SetStateAction } from 'react';
import React, { useContext, useEffect, useState } from 'react';
import {
  Button,
  ColumnAs,
  DataDataContext,
  Divider,
  EntityEditorDispatchContext,
  EntityEditorStateContext,
  EntityFieldEditor,
  Form,
  FormField,
  InputText,
  Loader,
  Message,
  Row,
} from '../..';
import type { DataDataContextValue, MessageItem } from '../..';
import type { EntityEditorDraftState, EntityEditorStateAction } from './EntityEditorReducer';
import {
  EntityUpsertedAction,
  ResetEntityAction,
  SetFieldAction,
  SetMessageLoadMessageAction,
  SetNameAction,
  UpdateEntityAction,
} from './EntityEditorReducer';

export interface EntityEditorProps {
  entityId: string;
}

interface EntityEditorInnerProps extends EntityEditorProps {
  draftState: EntityEditorDraftState;
  createEntity: DataDataContextValue['createEntity'];
  updateEntity: DataDataContextValue['updateEntity'];
}

export function EntityEditor({ entityId }: EntityEditorProps): JSX.Element | null {
  const editorState = useContext(EntityEditorStateContext);
  const dispatchEditorState = useContext(EntityEditorDispatchContext);

  const draftState = editorState.drafts.find((x) => x.id === entityId);
  if (!draftState) {
    throw new Error(`Can't find state for id (${entityId})`);
  }

  const { useEntity, createEntity, updateEntity } = useContext(DataDataContext);

  return (
    <>
      {draftState.exists ? (
        <EntityLoader {...{ entityId, useEntity, dispatchEditorState }} />
      ) : null}
      <EntityEditorInner
        {...{
          entityId,
          draftState,
          editorState,
          dispatchEditorState,
          createEntity,
          updateEntity,
        }}
      />
    </>
  );
}

export function EntityLoader({
  entityId,
  useEntity,
  dispatchEditorState,
}: {
  entityId: string;
  useEntity: DataDataContextValue['useEntity'];
  dispatchEditorState: Dispatch<EntityEditorStateAction>;
}): null {
  const { entity, entityError } = useEntity(entityId);

  useEffect(() => {
    if (entity) {
      dispatchEditorState(new UpdateEntityAction(entityId, entity));
    }
  }, [entityId, entity, dispatchEditorState]);

  useEffect(() => {
    dispatchEditorState(
      new SetMessageLoadMessageAction(
        entityId,
        entityError?.error ?? null,
        entityError
          ? {
              kind: 'danger',
              title: 'Failed loading entity',
              message: `${entityError.error}: ${entityError.message}`,
            }
          : null
      )
    );
  }, [dispatchEditorState, entityError, entityId]);

  return null;
}

function EntityEditorInner({
  entityId,
  draftState,
  createEntity,
  updateEntity,
}: EntityEditorInnerProps): JSX.Element {
  const editorState = useContext(EntityEditorStateContext);
  const dispatchEditorState = useContext(EntityEditorDispatchContext);
  const [submitLoading, setSubmitLoading] = useState(false);
  const [submitMessage, setSubmitMessage] = useState<MessageItem | null>(null);

  const { entity } = draftState;

  if (!entity) {
    if (draftState.initMessage) {
      return <Message {...draftState.initMessage} />;
    }
    if (draftState.entityLoadMessage) {
      return <Message {...draftState.entityLoadMessage} />;
    }
    return <Loader />;
  }

  const nameId = `${draftState.id}-_name`;

  return (
    <div id={entityId} data-entityid={entityId}>
      <ColumnAs
        as={Form}
        className="p-3 is-rounded has-background has-shadow"
        onSubmit={() =>
          submitEntity(
            draftState,
            setSubmitLoading,
            setSubmitMessage,
            createEntity,
            updateEntity,
            dispatchEditorState
          )
        }
        onReset={() => dispatchEditorState(new ResetEntityAction(entityId))}
      >
        {draftState.initMessage ? <Message {...draftState.initMessage} /> : null}
        {draftState.entityLoadMessage ? <Message {...draftState.entityLoadMessage} /> : null}
        <FormField htmlFor={nameId} label="Name">
          <InputText
            id={nameId}
            value={entity.name}
            onChange={(name) => dispatchEditorState(new SetNameAction(entityId, name))}
          />
        </FormField>
        <Divider />
        {entity.fields.map(({ fieldSpec, value }) => {
          const handleFieldChanged = (newValue: unknown) => {
            dispatchEditorState(new SetFieldAction(entityId, fieldSpec.name, newValue));
          };

          return (
            <EntityFieldEditor
              idPrefix={entityId}
              key={fieldSpec.name}
              schema={editorState.schema}
              fieldSpec={fieldSpec}
              value={value}
              onValueChanged={handleFieldChanged}
            />
          );
        })}

        <Row gap={1}>
          <Button kind="primary" type="submit" disabled={!entity.name} loading={submitLoading}>
            Save
          </Button>
          <Button type="reset" disabled={!entity}>
            Reset
          </Button>
        </Row>
        {submitMessage ? (
          <Message {...submitMessage} onDismiss={() => setSubmitMessage(null)} />
        ) : null}
      </ColumnAs>
    </div>
  );
}

function createAdminEntity(
  draftState: EntityEditorDraftState
): AdminEntityCreate | AdminEntityUpdate {
  const entityState = draftState.entity;
  if (!entityState) throw new Error('No entity in state');

  let result: AdminEntityCreate | AdminEntityUpdate;
  if (entityState.version === 0) {
    result = {
      id: draftState.id,
      _type: entityState.entitySpec.name,
      _name: entityState.name,
      _version: entityState.version,
    };
  } else {
    const { id } = draftState;
    if (!id) {
      throw new Error('Expected id');
    }
    result = { id, _type: entityState.entitySpec.name };
    if (entityState.name !== entityState.initialName) {
      result._name = entityState.name;
    }
  }
  for (const { fieldSpec, value, initialValue } of entityState.fields) {
    if (value !== initialValue) {
      result[fieldSpec.name] = value;
    }
  }
  return result;
}

async function submitEntity(
  draftState: EntityEditorDraftState,
  setSubmitLoading: Dispatch<SetStateAction<boolean>>,
  setSubmitMessage: Dispatch<SetStateAction<MessageItem | null>>,
  createEntity: DataDataContextValue['createEntity'],
  updateEntity: DataDataContextValue['updateEntity'],
  dispatchEditorState: Dispatch<EntityEditorStateAction>
) {
  setSubmitLoading(true);
  const entity = createAdminEntity(draftState);
  const isNew = entity._version === 0;
  const result = await (isNew
    ? createEntity(entity as AdminEntityCreate)
    : updateEntity(entity as AdminEntityUpdate));

  if (result.isOk()) {
    dispatchEditorState(new EntityUpsertedAction(draftState.id));
  } else {
    setSubmitMessage({
      kind: 'danger',
      title: isNew ? 'Failed creating entity' : 'Failed saving entity',
      message: `${result.error}: ${result.message}`,
    });
  }
  setSubmitLoading(false);
}
