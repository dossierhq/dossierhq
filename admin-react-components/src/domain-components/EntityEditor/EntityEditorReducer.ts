import type {
  AdminEntity,
  EntityTypeSpecification,
  FieldSpecification,
  Schema,
} from '@datadata/core';
import isEqual from 'lodash/isEqual';
import { v4 as uuidv4 } from 'uuid';
import type { MessageItem } from '../../generic-components/Message/Message';

export type EntityEditorSelector = { id: string } | { id?: string; newType: string };

export interface EntityEditorState {
  schema: Schema;
  drafts: EntityEditorDraftState[];
}

export interface EntityEditorDraftState {
  id: string;
  initMessage: MessageItem | null;
  entityLoadMessage: MessageItem | null;
  entity: null | {
    version: number;
    entitySpec: EntityTypeSpecification;
    name: string;
    initialName: string;
    fields: FieldEditorState[];
  };
}

interface FieldEditorState {
  fieldSpec: FieldSpecification;
  value: unknown;
  initialValue: unknown;
}

export interface EntityEditorStateAction {
  reduce(state: EntityEditorState): EntityEditorState;
}

export class AddEntityDraftAction implements EntityEditorStateAction {
  #entitySelector: EntityEditorSelector;
  constructor(entitySelector: EntityEditorSelector) {
    this.#entitySelector = entitySelector;
  }

  reduce(state: EntityEditorState): EntityEditorState {
    const id = this.#entitySelector.id ?? uuidv4();
    let message: MessageItem | null = null;
    let entity: EntityEditorDraftState['entity'] = null;
    if ('newType' in this.#entitySelector) {
      const type = this.#entitySelector.newType;
      const entitySpec = state.schema.getEntityTypeSpecification(type);
      if (entitySpec) {
        entity = createEditorEntityDraftState(entitySpec, null);
      } else {
        message = {
          kind: 'danger',
          message: `Can't create entity with unsupported type: ${type}`,
        };
      }
    }

    const draft: EntityEditorDraftState = {
      initMessage: message,
      entityLoadMessage: null,
      id,
      entity,
    };

    return { ...state, drafts: [...state.drafts, draft] };
  }
}

abstract class EntityEditorDraftStateAction implements EntityEditorStateAction {
  #id: string;

  constructor(id: string) {
    this.#id = id;
  }

  reduce(state: EntityEditorState): EntityEditorState {
    const draftIndex = state.drafts.findIndex((x) => x.id === this.#id);
    if (draftIndex < 0) {
      throw new Error(`Can't find draft for ${this.#id}`);
    }
    const draftState = state.drafts[draftIndex];
    const newDraftState = this.reduceDraft(draftState, state);
    if (draftState === newDraftState) {
      return state;
    }
    const newDrafts = [...state.drafts];
    newDrafts[draftIndex] = newDraftState;
    return { ...state, drafts: newDrafts };
  }

  abstract reduceDraft(
    draftState: EntityEditorDraftState,
    state: EntityEditorState
  ): EntityEditorDraftState;
}

export class SetMessageLoadMessageAction extends EntityEditorDraftStateAction {
  #message: MessageItem | null;

  constructor(id: string, message: MessageItem | null) {
    super(id);
    this.#message = message;
  }

  reduceDraft(
    draftState: EntityEditorDraftState,
    _state: EntityEditorState
  ): EntityEditorDraftState {
    if (isEqual(draftState.entityLoadMessage, this.#message)) {
      return draftState;
    }
    if (draftState.entity?.version === 0 && this.#message) {
      // Skip loading entity error for new entity
      return draftState;
    }
    return {
      ...draftState,
      entityLoadMessage: this.#message,
    };
  }
}

export class UpdateEntityAction extends EntityEditorDraftStateAction {
  #entity: AdminEntity;

  constructor(id: string, entity: AdminEntity) {
    super(id);
    this.#entity = entity;
  }

  reduceDraft(
    draftState: EntityEditorDraftState,
    state: EntityEditorState
  ): EntityEditorDraftState {
    //TODO handle update when there are local changes
    const entitySpec = state.schema.getEntityTypeSpecification(this.#entity._type);
    if (!entitySpec) {
      return {
        ...draftState,
        initMessage: {
          kind: 'danger',
          message: `Can't create entity with unsupported type: ${this.#entity._type}`,
        },
      };
    }

    return {
      ...draftState,
      entity: createEditorEntityDraftState(entitySpec, this.#entity),
    };
  }
}

export class SetNameAction extends EntityEditorDraftStateAction {
  #name: string;

  constructor(id: string, name: string) {
    super(id);
    this.#name = name;
  }

  reduceDraft(
    draftState: EntityEditorDraftState,
    _state: EntityEditorState
  ): EntityEditorDraftState {
    const { entity } = draftState;
    if (!entity) {
      throw new Error('Unexpected state, no entity');
    }
    return { ...draftState, entity: { ...entity, name: this.#name } };
  }
}

export class SetFieldAction extends EntityEditorDraftStateAction {
  #field: string;
  #value: unknown;

  constructor(id: string, field: string, value: unknown) {
    super(id);
    this.#field = field;
    this.#value = value;
  }

  reduceDraft(
    draftState: EntityEditorDraftState,
    _state: EntityEditorState
  ): EntityEditorDraftState {
    const { entity } = draftState;
    if (!entity) throw new Error('Unexpected state, no entity');
    const fields = [...entity.fields];
    const index = fields.findIndex((x) => x.fieldSpec.name === this.#field);
    if (index < 0) throw new Error(`Invalid field ${this.#field}`);
    fields[index] = { ...fields[index], value: this.#value };
    return { ...draftState, entity: { ...entity, fields } };
  }
}

export function reduceEntityEditorState(
  state: EntityEditorState,
  action: EntityEditorStateAction
): EntityEditorState {
  return action.reduce(state);
}

export function initializeEntityEditorState({ schema }: { schema: Schema }): EntityEditorState {
  return { schema, drafts: [] };
}

function createEditorEntityDraftState(
  entitySpec: EntityTypeSpecification,
  entity: AdminEntity | null
): EntityEditorDraftState['entity'] {
  const fields = entitySpec.fields.map((fieldSpec) => {
    const value = entity?.[fieldSpec.name] ?? null;
    return { fieldSpec, value, initialValue: value };
  });
  return {
    entitySpec,
    version: entity ? entity._version + 1 : 0,
    name: entity?._name ?? '',
    initialName: entity?._name ?? '',
    fields,
  };
}
