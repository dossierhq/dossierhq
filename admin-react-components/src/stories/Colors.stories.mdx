import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import { ColorSwatchGroup } from './ColorSwatch';

<Meta title="Colors" />

## Colors

<ColorSwatchGroup className="" />
<ColorSwatchGroup className="bg-blank1" />
<ColorSwatchGroup className="bg-blank2" />
<ColorSwatchGroup className="bg-primary" />
<ColorSwatchGroup className="bg-danger" />
