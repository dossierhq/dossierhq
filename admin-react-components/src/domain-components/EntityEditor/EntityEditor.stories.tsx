import type { Story } from '@storybook/react/types-6-0';
import React from 'react';
import { DataDataContext } from '../..';
import { EntityEditor } from './EntityEditor';
import type { EntityEditorProps } from './EntityEditor';
import { foo1Id, fooDeletedId, getEntityFixture } from '../../test/EntityFixtures';
import TestContextValue from '../../test/TestContextValue';

export default {
  title: 'Domain/EntityEditor',
  component: EntityEditor,
  argTypes: { onSubmit: { action: 'submit' } },
  args: {},
};

const Template: Story<EntityEditorProps> = (args) => {
  return (
    <DataDataContext.Provider value={new TestContextValue()}>
      <EntityEditor {...args} />
    </DataDataContext.Provider>
  );
};

export const NewFoo = Template.bind({});
NewFoo.args = { entity: { _type: 'Foo' }, idPrefix: 'new-entity-123' };

export const FullFoo = Template.bind({});
FullFoo.args = {
  entity: getEntityFixture(foo1Id),
};

export const DeletedFoo = Template.bind({});
DeletedFoo.args = {
  entity: getEntityFixture(fooDeletedId),
};
