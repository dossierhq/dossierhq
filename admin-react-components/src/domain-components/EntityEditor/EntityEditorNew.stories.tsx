import type { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import type { DataDataContextValue, EntityEditorNewProps } from '../..';
import { DataDataContext, EntityEditorNew } from '../..';
import type { EntityEditorSelector } from './EntityEditorReducer';
import { useEntityEditorState } from './EntityEditorReducer';
import { foo1Id, fooDeletedId } from '../../test/EntityFixtures';
import {
  createContextValue,
  SlowInterceptor,
  TestContextAdapter,
} from '../../test/TestContextAdapter';

type StoryProps = Omit<EntityEditorNewProps, 'editorState' | 'dispatchEditorState'> & {
  entitySelector: EntityEditorSelector;
  contextAdapter?: TestContextAdapter;
};

const meta: Meta<StoryProps> = {
  title: 'Domain/EntityEditorNew',
  component: EntityEditorNew,
  args: {},
};
export default meta;

const Template: Story<StoryProps> = (args) => {
  const contextValue = createContextValue(args?.contextAdapter);
  return (
    <DataDataContext.Provider value={contextValue}>
      <Wrapper entitySelector={args.entitySelector} contextValue={contextValue} />
    </DataDataContext.Provider>
  );
};

function Wrapper({
  entitySelector,
  contextValue,
}: {
  entitySelector: EntityEditorSelector;
  contextValue: DataDataContextValue;
}) {
  const { editorState, dispatchEditorState } = useEntityEditorState(entitySelector, contextValue);
  return <EntityEditorNew editorState={editorState} dispatchEditorState={dispatchEditorState} />;
}

export const NewFoo = Template.bind({});
NewFoo.args = { entitySelector: { id: '82ded109-44f2-48b9-a676-43162fda3d7d', newType: 'Foo' } };

export const FullFoo = Template.bind({});
FullFoo.args = { entitySelector: { id: foo1Id } };

export const DeletedFoo = Template.bind({});
DeletedFoo.args = { entitySelector: { id: fooDeletedId } };

export const SlowFullFoo = Template.bind({});
SlowFullFoo.args = {
  entitySelector: { id: foo1Id },
  contextAdapter: new TestContextAdapter(SlowInterceptor),
};

export const NotFound = Template.bind({});
NotFound.args = { entitySelector: { id: 'c6f97fae-1213-4be0-996f-4f20c7da7e65' } };

export const InvalidTypeNewEntity = Template.bind({});
InvalidTypeNewEntity.args = {
  entitySelector: { id: 'da4a48d6-341e-4515-941f-b83578191b51', newType: 'InvalidType' },
};
