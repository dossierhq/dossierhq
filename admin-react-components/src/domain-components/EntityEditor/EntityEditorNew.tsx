import type { AdminEntityCreate, AdminEntityUpdate } from '@datadata/core';
import type { Dispatch, SetStateAction } from 'react';
import React, { useContext, useState } from 'react';
import {
  Button,
  DataDataContext,
  Divider,
  EntityFieldEditor,
  Form,
  FormField,
  InputText,
  Loader,
  Message,
} from '../..';
import type { DataDataContextValue, MessageItem } from '../..';
import type { EntityEditorState, EntityEditorStateAction } from './EntityEditorReducer';
import { SetFieldAction, SetNameAction } from './EntityEditorReducer';

export interface EntityEditorNewProps {
  editorState: EntityEditorState;
  dispatchEditorState: Dispatch<EntityEditorStateAction>;
  style?: React.CSSProperties;
}

interface EntityEditorInnerProps {
  style?: React.CSSProperties;
  editorState: EntityEditorNewProps['editorState'];
  dispatchEditorState: Dispatch<EntityEditorStateAction>;
  createEntity: DataDataContextValue['createEntity'];
  updateEntity: DataDataContextValue['updateEntity'];
}

export function EntityEditorNew({
  editorState,
  dispatchEditorState,
  style,
}: EntityEditorNewProps): JSX.Element | null {
  const context = useContext(DataDataContext);
  if (!context) {
    return <Loader />;
  }

  const { createEntity, updateEntity } = context;

  return (
    <EntityEditorInner
      {...{
        editorState,
        dispatchEditorState,
        style,
        createEntity,
        updateEntity,
      }}
    />
  );
}

function EntityEditorInner({
  editorState,
  dispatchEditorState,
  style,
  createEntity,
  updateEntity,
}: EntityEditorInnerProps): JSX.Element {
  const [submitLoading, setSubmitLoading] = useState(false);
  const [submitMessage, setSubmitMessage] = useState<MessageItem | null>(null);

  const { entity } = editorState;

  if (!entity) {
    if (editorState.initMessage) {
      return <Message {...editorState.initMessage} />;
    }
    if (editorState.entityLoadMessage) {
      return <Message {...editorState.entityLoadMessage} />;
    }
    return <Loader />;
  }

  const nameId = `${editorState.id}-_name`;

  return (
    <Form
      onSubmit={() =>
        submitEntity(editorState, setSubmitLoading, setSubmitMessage, createEntity, updateEntity)
      }
      style={style}
    >
      {editorState.initMessage ? <Message {...editorState.initMessage} /> : null}
      {editorState.entityLoadMessage ? <Message {...editorState.entityLoadMessage} /> : null}
      <FormField htmlFor={nameId} label="Name">
        <InputText
          id={nameId}
          value={entity.name}
          onChange={(name) => dispatchEditorState(new SetNameAction(name))}
        />
      </FormField>
      <Divider />
      {entity.fields.map(({ fieldSpec, value }) => {
        const handleFieldChanged = (newValue: unknown) => {
          dispatchEditorState(new SetFieldAction(fieldSpec.name, newValue));
        };

        return (
          <EntityFieldEditor
            idPrefix={editorState.id}
            key={fieldSpec.name}
            schema={editorState.schema}
            fieldSpec={fieldSpec}
            value={value}
            onValueChanged={handleFieldChanged}
          />
        );
      })}

      <Button kind="primary" type="submit" disabled={!entity.name} loading={submitLoading}>
        Save
      </Button>
      {submitMessage ? (
        <Message {...submitMessage} onDismiss={() => setSubmitMessage(null)} />
      ) : null}
    </Form>
  );
}

function createAdminEntity(state: EntityEditorState): AdminEntityCreate | AdminEntityUpdate {
  const entityState = state.entity;
  if (!entityState) throw new Error('No entity in state');

  let result: AdminEntityCreate | AdminEntityUpdate;
  if (entityState.version === 0) {
    result = {
      _type: entityState.entitySpec.name,
      _name: entityState.name,
      _version: entityState.version,
    };
  } else {
    const { id } = state;
    if (!id) {
      throw new Error('Expected id');
    }
    result = { id, _type: entityState.entitySpec.name };
    if (entityState.name !== entityState.initialName) {
      result._name = entityState.name;
    }
  }
  for (const { fieldSpec, value, initialValue } of entityState.fields) {
    if (value !== initialValue) {
      result[fieldSpec.name] = value;
    }
  }
  return result;
}

async function submitEntity(
  editorState: EntityEditorState,
  setSubmitLoading: Dispatch<SetStateAction<boolean>>,
  setSubmitMessage: Dispatch<SetStateAction<MessageItem | null>>,
  createEntity: DataDataContextValue['createEntity'],
  updateEntity: DataDataContextValue['updateEntity']
) {
  setSubmitLoading(true);
  const entity = createAdminEntity(editorState);
  const isNew = entity._version === 0;
  const result = await (isNew
    ? createEntity(entity as AdminEntityCreate)
    : updateEntity(entity as AdminEntityUpdate));

  if (result.isError()) {
    setSubmitMessage({
      kind: 'danger',
      title: isNew ? 'Failed creating entity' : 'Failed saving entity',
      message: `${result.error}: ${result.message}`,
    });
  }
  setSubmitLoading(false);
}
