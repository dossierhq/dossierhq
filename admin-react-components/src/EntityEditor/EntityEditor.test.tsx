import type { Story } from '@storybook/react/types-6-0';
import { default as Default, DeletedFoo, FullFoo, NewFoo } from './EntityEditor.stories';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import type { EntityEditorProps } from './EntityEditor';

function renderStory(StoryUnderTest: Story<EntityEditorProps>) {
  const args = { ...Default.args, ...StoryUnderTest.args } as EntityEditorProps;
  const tree = renderer.create(<StoryUnderTest {...args} />).toJSON();
  return tree;
}

test('DeletedFoo', () => {
  expect(renderStory(DeletedFoo)).toMatchSnapshot();
});

test('FullFoo', () => {
  expect(renderStory(FullFoo)).toMatchSnapshot();
});

test('NewFoo', () => {
  expect(renderStory(NewFoo)).toMatchSnapshot();
});
