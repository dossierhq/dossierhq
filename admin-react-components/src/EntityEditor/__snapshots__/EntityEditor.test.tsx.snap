// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeletedFoo render 1`] = `
<form
  className="dd form has-background"
  onSubmit={[Function]}
>
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-_name"
    >
      Name
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-_name"
        onChange={[Function]}
        type="text"
        value="Hello"
      />
    </div>
  </div>
  <hr
    className="dd divider"
  />
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-title"
    >
      title
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <input
    className="dd button text-button bg-primary"
    disabled={false}
    type="submit"
    value="Save"
  />
</form>
`;

exports[`FullFoo render 1`] = `
<form
  className="dd form has-background"
  onSubmit={[Function]}
>
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-_name"
    >
      Name
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-_name"
        onChange={[Function]}
        type="text"
        value="Hello"
      />
    </div>
  </div>
  <hr
    className="dd divider"
  />
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-title"
    >
      title
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="entity-fc66b4d7-61ff-44d4-8f68-cb7f526df046-title"
        onChange={[Function]}
        type="text"
        value="Hello"
      />
    </div>
  </div>
  <input
    className="dd button text-button bg-primary"
    disabled={false}
    type="submit"
    value="Save"
  />
</form>
`;

exports[`NewFoo render 1`] = `
<form
  className="dd form has-background"
  onSubmit={[Function]}
>
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="new-entity-123-_name"
    >
      Name
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="new-entity-123-_name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <hr
    className="dd divider"
  />
  <div>
    <label
      className="dd text-subtitle1"
      htmlFor="new-entity-123-title"
    >
      title
    </label>
    <div
      className="dd form-control"
    >
      <input
        className="dd text-body1 input"
        id="new-entity-123-title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <input
    className="dd button text-button bg-primary"
    disabled={true}
    type="submit"
    value="Save"
  />
</form>
`;
